### PR TITLE
Arm setpoints & Sim

### PR DESCRIPTION
**Pull request by OliverW10**
Provides a way to tell the arm to go to set positions, also adds arm simulation and a rev encoder wrapper that is needed for the simulation

### CODE REVIEW

File: utilities/rev.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import wpilib
+
+
+class SparkMaxAbsoluteEncoderWrapper:
+    def __init__(self, motor: rev.CANSparkMax) -> None:
+        self.real_encoder = motor.getAbsoluteEncoder(
+            rev.SparkMaxAbsoluteEncoder.Type.kDutyCycle
+        )
+        self.sim_position = 0
+        self.sim_velocity = 0
```
Comment (LucienMorey) [inline utilities/rev.py:11]:
```suggestion
        self.sim_position = 0.0
        self.sim_velocity = 0.0
```
This should be the fix to the mypy stuff

File: components/arm.py [hunk 2]
```diff
@@ -49,25 +64,31 @@ def __init__(self) -> None:
         self._rotation_motor_right.follow(self.rotation_motor, invert=True)
         self._rotation_motor_right.setIdleMode(CANSparkMax.IdleMode.kCoast)
         self._rotation_motor_right.setInverted(False)
-
-        self.rotation_encoder = self.rotation_motor.getAbsoluteEncoder(
-            SparkMaxAbsoluteEncoder.Type.kDutyCycle
+        self.relative_encoder = SparkMaxEncoderWrapper(self.rotation_motor)
+        # ouput position is conversion factor * motor rotations
+        # should be < 1
+        self.relative_encoder.real_encoder.setPositionConversionFactor(
```
Comment (auscompgeek) [inline components/arm.py:70]: A facade shouldn't require breaking the abstraction. It should expose its own method.

File: components/arm.py [hunk 3]
```diff
@@ -83,76 +104,137 @@ def __init__(self) -> None:
...
-        is_currently_up = (
-            self.UPRIGHT_ANGLE > (self.get_angle() + math.pi / 2) > -self.UPRIGHT_ANGLE
+        self.chooser = SendableChooser()
+        self.chooser.addOption("Score Cone High", Setpoints.SCORE_CONE_HIGH)
+        self.chooser.addOption("Score Cube High", Setpoints.SCORE_CUBE_HIGH)
+        self.chooser.addOption("Score Cone Mid", Setpoints.SCORE_CONE_MID)
+        self.chooser.addOption("Score Cube mid", Setpoints.SCORE_CUBE_MID)
+        self.chooser.addOption("Handoff", Setpoints.HANDOFF)
+        self.chooser.setDefaultOption("Handoff", Setpoints.HANDOFF)
+        SmartDashboard.putData(self.chooser)
```
Comment (auscompgeek) [inline components/arm.py:122]: Where is the name for this chooser defined?

File: components/arm.py [hunk 4]
```diff
@@ -83,76 +104,137 @@ def __init__(self) -> None:
...
+            4,
+            wpilib.Color8Bit(68, 117, 26),
+        )
+        wpilib.SmartDashboard.putData("Arm sim", self.arm_mech2d)
+
+    def setup(self) -> None:
+        self.set_setpoint(Setpoints.SCORE_CONE_HIGH)
 
+    def execute(self) -> None:
+        setpoint = typing.cast(Setpoint, self.chooser.getSelected())
```
Comment (auscompgeek) [inline components/arm.py:157]: 😅 the type hints ought to allow this without casting. Could you please file an issue?

In the scenario where we're in the real robot control loop, it'd be preferable to type hint the variable and ignore the type error instead.

File: utilities/rev.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import wpilib
+
+
+class SparkMaxEncoderWrapper:
+    """Wraps an spark max relative encoder and allows you to set the velocity in the sim"""
+
+    def __init__(self, motor: rev.CANSparkMax) -> None:
+        self.real_encoder = motor.getEncoder()
+        self.sim_position = 0.0
+        self.sim_velocity = 0.0
```
Comment (auscompgeek) [inline utilities/rev.py:11]: Where's the sim support?

File: utilities/rev.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import rev
+import wpilib
+
+
+class SparkMaxEncoderWrapper:
+    """Wraps an spark max relative encoder and allows you to set the velocity in the sim"""
```
Comment (auscompgeek) [inline utilities/rev.py:6]: Why restrict this to relative encoders only?

**Review by auscompgeek — CHANGES_REQUESTED** [review]

File: components/arm.py (hunk 2 above)
Comment (auscompgeek) [reply components/arm.py:70]: Or perhaps indeed the (singular) conversion factor should be passed to the constructor, to guarantee the conversion factor is always set. IIRC we had that same design with the Talon encoder wrapper in the past.

File: components/arm.py [hunk 7]
```diff
@@ -172,63 +248,50 @@ def at_goal_extension(self, allowable_error=0.05) -> bool:
...
-    def on_enable(self):
+    def unbrake(self) -> None:
+        self.brake_solenoid.set(True)
+
+    def on_enable(self) -> None:
         self.extension_controller.reset(self.get_extension())
         self.rotation_controller.reset(self.get_angle())
+
+
+@dataclass
```
Comment (auscompgeek) [inline components/arm.py:265]: What dataclass features are being used here? I see our own `__init__` is being defined.

File: physics.py [hunk 8]
```diff
@@ -66,10 +74,57 @@ def __init__(self, physics_controller: PhysicsInterface, robot: MyRobot):
             for module in robot.chassis.modules
         ]
 
-        self.imu = wpilib.simulation.SimDeviceSim("navX-Sensor", 4)
+        self.arm = robot.arm
```
Comment (auscompgeek) [inline physics.py:77]: Please don't keep robot objects alive longer than the robot. Grab only the things needed instead, i.e. simulation objects.

File: components/arm.py [hunk 9]
```diff
@@ -172,63 +248,50 @@ def at_goal_extension(self, allowable_error=0.05) -> bool:
...
+        if angle > Arm.MAX_ANGLE:  # and (self.angle - math.tau) < Arm.MIN_ANGLE:
+            angle -= math.tau
+        self.angle = clamp(angle, Arm.MIN_ANGLE, Arm.MAX_ANGLE)
+
+        if self.angle != angle or self.extension != extension:
+            print(
+                "SETPOINT WAS CLAMPED", (angle, extension), (self.angle, self.extension)
+            )
+
+    @staticmethod
```
Comment (auscompgeek) [inline components/arm.py:285]: Use a classmethod instead.

File: components/arm.py [hunk 10]
```diff
@@ -161,7 +233,11 @@ def set_length(self, value: float) -> None:
         """Sets a goal length to go to in meters"""
         self.goal_extension = clamp(value, self.MIN_EXTENSION, self.MAX_EXTENSION)
 
-    DEFAULT_ALLOWABLE_ANGLE_ERROR = math.radians(5)
+    def set_setpoint(self, value: "Setpoint") -> None:
```
Comment (auscompgeek) [inline components/arm.py:236]: nit: Can you move the Setpoint class higher up to avoid forward references?

Comment (OliverW10) [reply components/arm.py:236]: the setpoints class needs the arm for its MIN/MAX_EXTENSION/ANGLE and i thought the Setpoint and Setpoints classes would be better next to each other

Comment (auscompgeek) [reply components/arm.py:236]: Would it make sense to pull those constants out of the Arm class?

File: physics.py (hunk 8 above)
Comment (OliverW10) [reply physics.py]: is grabbing things like motors and solenoids to .get their output ok?

Comment (auscompgeek) [reply physics.py]: No. Those should be available through the simulation backend as well.

**Review by whert-dev — APPROVED** [review]

**Pull request merged**